controllers/clustertopology_controller_sync.go

```diff
@@ -16,10 +16,68 @@ limitations under the License.
 
 package controllers
 
-import "context"
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
 
 // reconcileState reconciles the current and desired state of the managed Cluster topology.
 func (r *ClusterTopologyReconciler) reconcileState(ctx context.Context, current, desired *clusterTopologyState) error {
-	// TODO: add reconcile logic
 	return nil
 }
+
+type twoWaysMergePatchHelper struct {
+	client   client.Client
+	original client.Object
+	data     []byte
+}
+
+// createTwoWaysMergePatchHelper will return a patch that yields the modified document when applied to the original document.
+// NOTE: In the case of ClusterTopologyReconciler, original is the current object, modified is the desired object, and
+// the patch returns all the changed required to align current to what is defined in desired; fields not defined in desired
+// are going to be preserved without changes.
+func createTwoWaysMergePatchHelper(original, modified client.Object, crClient client.Client) (*twoWaysMergePatchHelper, error) {
+	originalJSON, err := json.Marshal(original)
+	if err != nil {
+		return nil, err
+	}
+
+	modifiedJSON, err := json.Marshal(modified)
+	if err != nil {
+		return nil, err
+	}
+
+	originalWithModifiedJSON, err := jsonpatch.MergePatch(originalJSON, modifiedJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := jsonpatch.CreateMergePatch(originalJSON, originalWithModifiedJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	return &twoWaysMergePatchHelper{
+		client:   crClient,
+		data:     data,
+		original: original,
+	}, nil
+}
+
+// HasChanges return true if the patch has some changes.
+func (h *twoWaysMergePatchHelper) HasChanges() bool {
```

Comment:
I'm not sure if we should export HasChanges and Patch while createTwoWaysMergePatchHelper and twoWaysMergePatchHelper are not exported

Reply:
This does not really matter given that the object is private
However, I think it is a good practice to keep the methods that the consumers are expected to use as public, and the internals of the object as private; this will make simpler and straight forward moving the object to another package in the future.